src/lib/rsa.c

```diff
@@ -294,180 +294,47 @@ pgp_rsa_decrypt_pkcs1(uint8_t *               out,
-    pgp_push_checksum_writer(output, seckey);
-
-    if ((seckey->pubkey.alg != PGP_PKA_RSA) &&
-        (seckey->pubkey.alg != PGP_PKA_RSA_ENCRYPT_ONLY) &&
-        (seckey->pubkey.alg != PGP_PKA_RSA_SIGN_ONLY)) {
-        RNP_LOG("Bad seckey->pubkey.alg");
-        goto end;
-    }
-
-    CHECK(pgp_write_mpi(output, seckey->key.rsa.d), true, false);
-    CHECK(pgp_write_mpi(output, seckey->key.rsa.p), true, false);
-    CHECK(pgp_write_mpi(output, seckey->key.rsa.q), true, false);
-    CHECK(pgp_write_mpi(output, seckey->key.rsa.u), true, false);
-    ret = true;
-
-end:
-    if (!ret) {
-        botan_mp_destroy(rsa_n);
-        botan_mp_destroy(rsa_e);
-        botan_mp_destroy(rsa_d);
-        botan_mp_destroy(rsa_p);
-        botan_mp_destroy(rsa_q);
-        botan_mp_destroy(rsa_u);
-
-        destroy_BN_mp(&seckey->pubkey.key.rsa.n);
-        destroy_BN_mp(&seckey->pubkey.key.rsa.e);
-        destroy_BN_mp(&seckey->key.rsa.d);
-        destroy_BN_mp(&seckey->key.rsa.p);
-        destroy_BN_mp(&seckey->key.rsa.q);
-        destroy_BN_mp(&seckey->key.rsa.u);
-    }
-
-    pgp_teardown_memory_write(output, mem);
-    botan_privkey_destroy(rsa_key);
-    botan_rng_destroy(rng);
-    return ret;
-}
-
-/**
- \ingroup HighLevel_KeyGenerate
- \brief Creates a self-signed RSA keypair
- \param numbits Modulus size
- \param e Public Exponent
- \param userid User ID
- \return The new keypair or NULL
-
- \note It is the caller's responsibility to call pgp_keydata_free(keydata)
- \sa rsa_generate_keypair()
- \sa pgp_keydata_free()
-*/
-pgp_key_t *
-pgp_rsa_new_selfsign_key(const int           numbits,
-                         const unsigned long e,
-                         uint8_t *           userid,
-                         const char *        hashalg,
-                         const char *        cipher)
-{
-    pgp_key_t *keydata;
-
-    keydata = pgp_keydata_new();
-    if (!rsa_generate_keypair(keydata, numbits, e, hashalg, cipher) ||
-        !pgp_add_selfsigned_userid(keydata, userid)) {
-        pgp_keydata_free(keydata);
-        return NULL;
-    }
-    return keydata;
-}
-
-pgp_key_t *
-pgp_rsa_new_key(const int           numbits,
-                const unsigned long e,
-                const char *        hashalg,
-                const char *        cipher)
-{
-    pgp_key_t *keydata = NULL;
-
-    keydata = pgp_keydata_new();
-    if (!rsa_generate_keypair(keydata, numbits, e, hashalg, cipher)) {
-        pgp_keydata_free(keydata);
-        return NULL;
-    }
-    return keydata;
-}
+int pgp_genkey_rsa(pgp_seckey_t* seckey, size_t numbits)
+   {
+   botan_privkey_t rsa_key = NULL;
+   botan_rng_t     rng = NULL;
+   int ret = 0;
+
+   if (botan_rng_init(&rng, NULL) != 0)
+      goto end;
+
+   if (botan_privkey_create_rsa(&rsa_key, rng, numbits) != 0)
+      goto end;
+
+   if (botan_privkey_check_key(rsa_key, rng, 1) != 0)
+      goto end;
+
+   /* Calls below never fail as calls above were OK */
```

Comment:
I believe it's a sanity check to ensure the key was generated with valid values. It probably does slow things down though, esp. with the thorough flag here.